### PR TITLE
Refactor @truffle/compile-solidity's CompilerSupplier

### PR DESCRIPTION
**Pull request by gnidan**
This PR refactors the CompilerSupplier and its use in @truffle/compile-solidity.

## History of this work

This PR was intended to included additional behavior to abstract the notion of a "Supplier", available for use by other languages and even by other kinds of components. **This PR itself has represented a feature branch with the following approved+merged PRs already included**:

- #4222 to clean up the existing interface (and the implementation, to a lesser extent)
- #4223 to turn on TypeScript in @truffle/compile-solidity so it's available in subsequent PRs

For reference, the remaining work originally was intended to include:
- #4224 to define an abstract supplier and contains the code to link strategies together.
- #4225 to update @truffle/compile-solidity's CompilerSupplier to use the new abstraction

Since those latter two PRs introduced breaking changes of their own, the thought was to wait for those before releasing this.

## Current status

At this point, the refactor that this PR already introduces (from the former two PRs mentioned) would be rather useful to some other efforts, e.g. to address the concern brought up in issue #4384. Since the latter two PRs above have no clear path forward, I propose that we can stomach a second breaking change if and when we want to continue that effort, and that we should introduce these already approved changes into `develop` much sooner than that!

### CODE REVIEW

File: .github/workflows/nodejs.yml
Comment (gnidan) [inline .github/workflows/nodejs.yml:10]: **Remove this before merge**

File: .github/workflows/nodejs.yml
Comment (gnidan) [inline .github/workflows/nodejs.yml:10]: **Remove this before merge**

Comment (Tuananh11021986) [reply .github/workflows/nodejs.yml:10]: Mình phải làm gì, ai chỉ mình với , mình người mới ko biết gì

**Comment by haltman-at** [comment]
Oh, hey, another note on `compilerRoots`.  Remember that comment about how we ought to change `https://ethereum.github.io/solc-bin/bin/` to `https://binaries.soliditylang.org/bin/`?

Well, also, while we should definitely include `https://binaries.soliditylang.org/bin/`, we should also probably include above it `https://binaries.soliditylang.org/emscripent-wasm32/`; this only includes released versions, not prereleases, but they're supposed to be faster than the ones in `/bin`. :)

**Review by haltman-at — DISMISSED** [review]
Well, I already approved the contents of this PR, so I may as well approve this, right?

I do have two comments:
1. In reviewing the PRs that went into this, there were some things I pointed out that got put off for later and filed as issues.  Those issues are still open; and that if the other work that was supposed to happen on this branch is being shelved for now, those issues should still be fixed at some point, hopefully soon.
2. I never got to reading the final PR that was supposed to go into this, but it had actual functional changes, right?  What's going to happen with those?  Or am I misremembering?

**Comment by gnidan** [comment]
> In reviewing the PRs that went into this, there were some things I pointed out that got put off for later and filed as issues. Those issues are still open; and that if the other work that was supposed to happen on this branch is being shelved for now, those issues should still be fixed at some point, hopefully soon.

Let's aggregate those! I'll post this comment and then I'll go and see if I can make separate comments on the "Files changed" tab so we can address these systematically.

- https://github.com/trufflesuite/trufflesuite.com/issues/1043
- https://github.com/trufflesuite/truffle/pull/4222#discussion_r687275690
- https://github.com/trufflesuite/truffle/pull/4222#discussion_r688150554
- https://github.com/trufflesuite/truffle/pull/4225/files#r688150977
- https://github.com/trufflesuite/truffle/pull/4225#discussion_r688148594
- https://github.com/trufflesuite/truffle/pull/4222#discussion_r687283884
- https://github.com/trufflesuite/truffle/pull/4222#discussion_r687287842
- https://github.com/trufflesuite/truffle/issues/4261
- https://github.com/trufflesuite/truffle/pull/4222#discussion_r687297120
- https://github.com/trufflesuite/truffle/pull/4222#issuecomment-898062803

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/index.ts:17]: We opened [this documentation issue](https://github.com/trufflesuite/trufflesuite.com/issues/1043) to ensure this knowledge is available to our users.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/index.ts:41]: @haltman-at writes [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687275690) about code duplication, which would've been addressed systematically in #4225.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/index.ts:93]: @haltman-at mentions [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687275987) that this doesn't get an Error class

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/index.ts:103]: @haltman-at added a note [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687276676) about this

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/index.ts:114]: @haltman-at notes [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687277390) that this method name is terrible.

File: packages/core/lib/commands/compile.js
Comment (gnidan) [inline packages/core/lib/commands/compile.js:117]: @haltman-at asks about Local/Native strategies [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687283884)

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:31]: @cameel brought [this](https://github.com/trufflesuite/truffle/pull/4222#discussion_r689001206) to our attention and we need to make sure we've got this information properly processed.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/index.ts:9]: @haltman-at suggested [this issue](https://github.com/trufflesuite/truffle/issues/4261), identifying that this CompilerSupplier implementation does too much manual retry logic, rather than using something like axios-retry.

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts:93]: @haltman-at writes [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687297120) how it's confusing that these are AsyncIterables, which warrants better commenting.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/index.ts:32]: @haltman-at writes [here](https://github.com/trufflesuite/truffle/pull/4222#pullrequestreview-728040585) that this is likely a mistake.

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:64]: @haltman-at notes [here](https://github.com/trufflesuite/truffle/pull/4222#pullrequestreview-728040585) that we may be able to make use of the semver package's `maxSatisfying` function

**Comment by gnidan** [comment]
> I never got to reading the final PR that was supposed to go into this, but it had actual functional changes, right? What's going to happen with those? Or am I misremembering?

@haltman-at that final PR had one or two minor functional changes, but I don't think that should get in the way of this PR, which is just cleanup, really. I can re-open that final PR to make sure we don't forget to do some follow-up there, if you'd like.

**Comment by haltman-at** [comment]
OK, I guess I'm dismissing my approval until we figure out what we're doing about the above stuff!  (Merge without them, then do them as follow-on work?  Do them on this branch?  Do them on another branch that gets merged into this?)

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/index.ts:41]: I posit that this duplication is not bad enough to warrant addressing at this time.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/index.ts:9]: I posit that, since there's an open issue, this needn't block merge of this PR.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/index.ts:17]: I posit that, since there's an open issue, this needn't block merge of this PR.

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (gnidan) [inline packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:13]: @haltman-at notes [here](https://github.com/trufflesuite/truffle/pull/4222#pullrequestreview-728040585) that we may be able to make use of the semver package's `maxSatisfying` function, eliminating this class's `findNewestValidVersion` method.

Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:13]: Looking to make this change, these behaviors are rather different - `maxSatisfying` takes a list of version, but our `findNewestValidVersion` takes this structured data of `{ releases, prereleases, builds }` that is nontrivial to dissect and clean up.

I posit that it's not worth the risk + effort to do this right now.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: (Pushed fix)

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts:93]: Noted a comment [here](https://github.com/trufflesuite/truffle/pull/4262#diff-94e2cd9540d4f8bd22a774f9e893ef938ff57e71e4b18d2a4f2c51f79f48a83fR73-R82)

File: packages/core/lib/commands/compile.js
Comment (gnidan) [reply packages/core/lib/commands/compile.js:117]: Turns out we already addressed this in #4222

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:31]: Seems like the entire comment suffices as an issue, which I've opened [here](https://github.com/trufflesuite/truffle/issues/4425)

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: (I've inlined this)

**Comment by gnidan** [comment]
I just observed that this PR wasn't working with `version: "pragma"`, so I've added a [commit](https://github.com/trufflesuite/truffle/pull/4262/commits/c43dfe86d9ffbcdd3d0e8b2f535e0f3e38e02c26) to address that.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/index.ts:17]: Makes sense, that issue's on a separate repo anyway.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: I'm not seeing the fix?

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:13]: The only information from that structure that it uses, however, is `Object.keys(releases)`; it discards the rest.  So I don't think it's so far off as you're suggesting.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: Hm, this raises another issue though -- like `spawn` and `compilerRoots`, the `dockerTagsUrl` option is undocumented.  I'll go comment on the existing issue to note that.

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: Ah, I see you've changed this. :)

File: packages/compile-solidity/src/compilerSupplier/index.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/index.ts]: Oh, you've changed this too. :)

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (gnidan) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:13]: Alright, I switched to using `maxSatisfying`, but it wasn't clean enough to warrant inlining, so I've kept `findNewestValidVersion`

**Comment by gnidan** [comment]
@haltman-at looks like I got everything outstanding and fixed those CI issues... this is back to you! Thanks!

File: packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts
Comment (haltman-at) [reply packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts:13]: OK, makes sense.

**Review by haltman-at — APPROVED** [review]
Alright, seems good to me!  For real this time I think. :)

**Comment by gnidan** [comment]
Thank you! Merging!

**Pull request merged**
